### PR TITLE
shell-ui : Add users group mapping in shell UI config

### PR DESCRIPTION
**Pull request by JBWatenbergScality**
**Component**:

shell-ui

**Context**: 

In order to disable features in Metalk8s UI and Solutions UI according to user permissions we need an abstract way to define user groups independant from any OIDC provider capabilities.

**Summary**:

This PR add support for an `userGroupsMapping` property in shell ui configuration that allows static user groups configuration.

**Acceptance criteria**: 

Shell-UI should expose aggregated users groups to Metalk8s and solutions UIs.

Refs : #3159

Depends on : https://github.com/scality/metalk8s/pull/3154 and https://github.com/scality/metalk8s/pull/3181

### CODE REVIEW

**Comment by JBWatenbergScality** [comment]
@bert-e after_pull_request=3181

File: shell-ui/src/auth/permissionUtils.spec.js [hunk 1]
```diff
@@ -116,3 +117,21 @@ describe('permission utils - isPathAccessible', () => {
...
+    expect(groups).toEqual(oidcGroups);
+  });
+
+  it('should return a merged array of groups when defined in OIDC claims and mapping ', () => {
+    //E
+    const oidcAndStaticGroups = ["group"];
+    const groups = getUserGroups({profile: {email: "test@test.com", groups: oidcAndStaticGroups}}, {'test@test.com': oidcAndStaticGroups});
+    //V
+    expect(groups).toEqual(oidcAndStaticGroups);
+  });
```
Comment (gdemonet) [inline shell-ui/src/auth/permissionUtils.spec.js]: Maybe this test case could be improved a bit, by making sure the input OIDC and mapped groups are different (otherwise we're not really testing the "merge" behaviour):
- have one case where OIDC groups are not set at all (to emulate Dex :wink:)
- have another case where there are some duplicates which get discarded by this method

File: shell-ui/src/index.js [hunk 2]
```diff
@@ -21,6 +21,8 @@ export type MenuItems = {[path: string]: TranslationAndGroups }
 
 export type Options = { main: MenuItems, subLogin: MenuItems };
 
+export type UserGroupsMapping = {[email: string]: string[]};
```
Comment (gdemonet) [inline shell-ui/src/index.js:24]: Static mapping is fine for now I guess, maybe adding support for regular expressions could be useful in the future.

Anyway. This mapping would need to be defined from Salt (could be merged with user-provided values in UI CSC) by reading the Dex CSC, so we can always get the "platform-admin" (or whatever it's called) group to Dex static users.

**Review by gdemonet — COMMENTED** [review]
LGTM :ok_hand: you may want to address usage of this new configuration in the same PR (defining a mapping for Dex users).

File: shell-ui/src/index.js (hunk 2 above)
Comment (gdemonet) [reply shell-ui/src/index.js:24]: In practice, I think this simply means updating the `salt/metalk8s/addons/ui/config/shell-ui-config.yaml.j2` template to:

```jinja
#!jinja|yaml


{%- set dex_defaults = salt.slsutil.renderer('salt://metalk8s/addons/dex/config/dex.yaml.j2', saltenv=saltenv) %}
{%- set dex = salt.metalk8s_service_configuration.get_service_conf('metalk8s-auth', 'metalk8s-dex-config', dex_defaults) %}

# Defaults for shell UI configuration
apiVersion: addons.metalk8s.scality.com/v1alpha1
kind: ShellUIConfig
spec:
  oidc:
    providerUrl: "/oidc"
    redirectUrl: "https://{{ grains.metalk8s.control_plane_ip }}:8443/"
    clientId: "metalk8s-ui"
    responseType: "id_token"
    scopes: "openid profile email groups offline_access audience:server:client_id:oidc-auth-client"
  userGroupsMapping:
{%- for user in dex.spec.config.staticPasswords | map(attribute='email') %}
    {{ user }}: [admin]
{%- endfor %}
  options:
    main:
      "https://{{ grains.metalk8s.control_plane_ip }}:8443/":
        en: "Platform"
        fr: "Plateforme"
      "https://{{ grains.metalk8s.control_plane_ip }}:8443/alerts":
        en: "Alerts"
        fr: "Alertes"
    subLogin:
      "https://{{ grains.metalk8s.control_plane_ip }}:8443/docs":
        en: "Documentation"
        fr: "Documentation"
```

File: docs/operation/account_administration/user_authentication_and_identity_management.rst [hunk 3]
```diff
@@ -103,6 +103,21 @@ the following steps from the bootstrap node.
...
+                         metalk8s.addons.ui.deployed saltenv=metalk8s-|version|
+
+   .. parsed-literal::
+
+      root\@bootstrap $ kubectl exec -n kube-system -c salt-master \\
+                         --kubeconfig /etc/kubernetes/admin.conf \\
+                         salt-master-bootstrap -- salt-run state.sls \\
+                         metalk8s.addons.prometheus-operator.deployed \\
+                         saltenv=metalk8s-|version|
+
```
Comment (gdemonet) [inline docs/operation/account_administration/user_authentication_and_identity_management.rst]: You can pass multiple SLS formulas to [`state.sls`](https://docs.saltproject.io/en/latest/ref/modules/all/salt.modules.state.html#salt.modules.state.sls), so the following should be enough:

```rst
   .. parsed-literal::

      root\@bootstrap $ STATES=$(printf ",metalk8s.addons.%s.deployed" \\
                                 dex prometheus-operator ui)
      root\@bootstrap $ kubectl exec -n kube-system -c salt-master \\
                         --kubeconfig /etc/kubernetes/admin.conf \\
                         salt-master-bootstrap -- salt-run state.sls \\
                         "${STATES:1}" saltenv=metalk8s-|version|
```

File: docs/operation/cluster_and_service_configuration.rst [hunk 4]
```diff
@@ -127,9 +127,31 @@ Features exposed include:
 
 The default Shell UI configuration values are specified below:
 
-.. literalinclude:: ../../salt/metalk8s/addons/ui/config/metalk8s-shell-ui-config.yaml.j2
-   :language: yaml
-   :lines: 3-
```
Comment (gdemonet) [inline docs/operation/cluster_and_service_configuration.rst]: Why not keep this include? It's easier to keep things in-sync this way IMO.

I think keeping Jinja syntax in there isn't really a problem (it's somewhat readable), but if it is, we could use the `.. jinja::` directive to render some of the templated values (such as the `userGroupsMapping`). Not really straightforward though, since I guess we want to keep a placeholder for the control plane IP.

File: salt/metalk8s/addons/ui/config/metalk8s-shell-ui-config.yaml.j2 [hunk 5]
```diff
@@ -11,8 +14,9 @@ spec:
     responseType: "id_token"
     scopes: "openid profile email groups offline_access audience:server:client_id:oidc-auth-client"
   userGroupsMapping:
-    "admin@metalk8s.invalid":
-      - admin
+{%- for user in dex.spec.config.staticPasswords | map(attribute='email') %}
+    "{{ user }}": [admin]
```
Comment (gdemonet) [inline salt/metalk8s/addons/ui/config/metalk8s-shell-ui-config.yaml.j2]: So, IIUC, in practice all Dex static users have the `admin` group mapped. Do we want another PR to filter MetalK8s pages based on this group? Also, to keep things clean, I would name this group `metalk8s:admin` (it's configurable anyway).

**Review by gdemonet — CHANGES_REQUESTED** [review]

File: docs/operation/cluster_and_service_configuration.rst (hunk 4 above)
Comment (JBWatenbergScality) [reply docs/operation/cluster_and_service_configuration.rst]: I had some doc build issues linked to yaml highlighting. Maybe using `.. jinja::` directive could help solves this, I will try it.

Comment (JBWatenbergScality) [reply docs/operation/cluster_and_service_configuration.rst]: I think I will remove highlighting for now WDYT ?

**Review by gdemonet — APPROVED** [review]
